File: src/main/java/de/bst/five_cities/opengeodb/OpenGeoDB.java
package de.bst.five_cities.opengeodb;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.util.Map;
import java.util.NoSuchElementException;
import java.util.TreeMap;

import org.apache.commons.io.FileUtils;
import org.apache.commons.io.LineIterator;

import de.bst.five_cities.utils.DownloadUtils;

/**
 * @see http://opengeodb.org/wiki/OpenGeoDB
 *
 */
public class OpenGeoDB {

	public static final String DOWNLOAD_BASE_URL = "http://www.fa-technik.adfc.de/code/opengeodb";
	public static final String LIECHTENSTEIN_TSV = "LI.tab";
	private static File localDir = new File("tmp/opengeodb");

	public static File getTSVFile(String countryID) {
		try {
			return DownloadUtils.getFile(new URL(DOWNLOAD_BASE_URL + "/" + countryID + ".tab"), localDir);
		} catch (Exception e) {
			throw new RuntimeException("Error downloading country file for " + countryID, e);
		}
	}

	public static Map<String, Integer> getTSVColumnMapping(String countryID) {

		Map<String, Integer> result = new TreeMap<>();
		File tsvFile = getTSVFile(countryID);
		String firstLine = firstLine(tsvFile);
		if (firstLine.charAt(0) != '#')
			throw new NoSuchElementException("Missing Header line in " + tsvFile);

		String[] splitted = firstLine.substring(1).split("\t");
		for (int i = 0; i < splitted.length; i++)
			result.put(splitted[i], i);
		return result;

	}

	private static String firstLine(File tsv) {
		try {
			LineIterator lineIterator = FileUtils.lineIterator(tsv);
			String firstLine = lineIterator.next();
			lineIterator.close();
			return firstLine;
		} catch (IOException e) {
			throw new RuntimeException("Error reading country file for " + tsv, e);
		}
	}

}

File: src/main/scala/de/bst/five_cities/opengeodb/OpenGeoDB.scala
package de.bst.five_cities.opengeodb

import java.io.File
import java.net.URL

import de.bst.five_cities.utils.DownloadUtils

import scala.io.Source

/**
  * @see http://opengeodb.org/wiki/OpenGeoDB
  */
object OpenGeoDB {
  private val DOWNLOAD_BASE_URL = "http://www.fa-technik.adfc.de/code/opengeodb"
  private val LOCAL_DIR = new File("tmp/opengeodb")

  def getTSVFile(countryID: String): File =
    DownloadUtils getFile(new URL(s"$DOWNLOAD_BASE_URL/$countryID.tab"), LOCAL_DIR)

  def getTSVColumnMapping(countryID: String): Map[String, Int] = {
    val tsvFile: File = getTSVFile(countryID)
    val firstLine = getFirstLine(tsvFile)
    if (!firstLine.startsWith("#")) throw new NoSuchElementException(s"Missing Header line in $tsvFile")
    (firstLine substring 1 split "\t").view.zipWithIndex.toMap
  }

  private def getFirstLine(tsv: File): String = (Source fromFile tsv).getLines.next
}
